hydro_lang/src/sim/runtime.rs

```diff
@@ -418,3 +437,161 @@ impl<T: Clone> SimHook for SingletonHook<T> {
         }
     }
 }
+
+pub struct KeyedSingletonHook<K: Hash + Eq + Clone, V: Clone> {
+    input: Rc<RefCell<FxHashMap<K, VecDeque<V>>>>, // FxHasher is deterministic
+    to_release: Option<Vec<(K, V, bool)>>,         // (key, data, is new)
+    last_released: FxHashMap<K, V>,
+    skipped_states: FxHashMap<K, Vec<V>>,
+    output: UnboundedSender<(K, V)>,
+    batch_location: (&'static str, &'static str, &'static str),
+    format_key_debug: fn(&K) -> Option<String>,
+    format_value_debug: fn(&V) -> Option<String>,
```

Comment:
These all really need docs
And prefer `/// this is docs for the next line` format